### PR TITLE
OJ-2606: Changed StrategyMapping param to a json string to support multiple st…

### PR DESCRIPTION
**Pull request by mikebeeby**
…rategies

### What changed

- Changed IIQStrategy to JSON to enable adding more types.
- Add score param to question handler to enable using the score to request different question strategies. 
- Updated tests in the QuestionHandler for the above.

### Why did it change

As a developer,
I want to ensure the Experian KBV API strategy mapping supports multiple strategies,
so that it can handle both the existing Medium Confidence journey and the new Low Confidence journey.

### Issue tracking

- [OJ-2606](https://govukverify.atlassian.net/browse/OJ-2606)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

[OJ-2606]: https://govukverify.atlassian.net/browse/OJ-2606?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/question/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionHandler.java [hunk 1]
```diff
@@ -236,11 +241,23 @@ private QuestionsResponse getQuestionAnswerResponse(
         return this.kbvService.getQuestions(questionRequest);
     }
 
-    private QuestionRequest createQuestionRequest(PersonIdentityDetailed personIdentity) {
+    private QuestionRequest createQuestionRequest(PersonIdentityDetailed personIdentity)
+            throws JsonProcessingException, InvalidStrategyScoreException {
         var questionRequest = new QuestionRequest();
-        var strategy = this.configurationService.getParameterValue(IIQ_STRATEGY_PARAM_NAME);
+        var strengthScoreParam =
```
Comment (jkunle) [inline lambdas/question/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionHandler.java:247]: I think this should be the verificationScore not strenghtscore

File: lambdas/question/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionHandler.java [hunk 2]
```diff
@@ -236,11 +241,23 @@ private QuestionsResponse getQuestionAnswerResponse(
...
+        var strategyParam = this.configurationService.getParameterValue(IIQ_STRATEGY_PARAM_NAME);
 
+        Map<String, String> strategyMap =
+                objectMapper.readValue(strategyParam, new TypeReference<>() {});
+        String strategy = strategyMap.get(strengthScoreParam);
+        LOGGER.info("Using IIQStrategy: {}", strategy);
+        if (strategy == null) {
+            throw new InvalidStrategyScoreException(
+                    "No question strategy found for score provided");
+        }
```
Comment (jkunle) [inline lambdas/question/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionHandler.java:258]: Could this be in some method

File: infrastructure/lambda/template.yaml [hunk 3]
```diff
@@ -139,11 +139,11 @@ Mappings:
...
-      dev: 3 out of 4 prioritised
-      build: 3 out of 4 prioritised
-      staging: 3 out of 4 prioritised
-      integration: 3 out of 4 prioritised
-      production: 3 out of 4 prioritised
+      dev: "{\"2\": \"3 out of 4 prioritised\"}"
+      build: "{\"2\": \"3 out of 4 prioritised\"}"
+      staging: "{\"2\": \"3 out of 4 prioritised\"}"
+      integration: "{\"2\": \"3 out of 4 prioritised\"}"
+      production: "{\"2\": \"3 out of 4 prioritised\"}"
```
Comment (CharlesIC) [inline infrastructure/lambda/template.yaml:146]: Can we please use the [Fn::ToJsonString function](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-ToJsonString.html) from the [AWS::LanguageExtensions transform](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-languageextension-transform.html) to encode YAML objects as JSON strings and avoid escaping.

The language extenions transform must come before the `AWS::Serverless` transform in the list at the top of the template.

A usage example in one of our templates [here](https://github.com/govuk-one-login/ipv-cri-check-hmrc-api/blob/652ee49acb4d0dfa2c5a03403358ee914a424e4f/infrastructure/template.yaml#L869-L871)

File: lambdas/question/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionHandler.java [hunk 4]
```diff
@@ -57,6 +59,7 @@ public class QuestionHandler
     public static final String HEADER_SESSION_ID = "session-id";
     public static final String LAMBDA_NAME = "get_question";
     public static final String ERROR_KEY = "error";
+    public static final String VERIFICATION_SCORE_PARAM_NAME = "di-ipv-cri-kbv-api/strengthScore";
```
Comment (jkunle) [inline lambdas/question/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionHandler.java:62]: `di-ipv-cri-kbv-api/strengthScore` is currently strengthSore see https://github.com/govuk-one-login/architecture/blob/83e2d11898edcd82edaccf17bc29ce6c403b5fbf/rfc/0024-identity-vc-for-checks.md#evidence-properties

File: infrastructure/lambda/template.yaml (hunk 3 above)
Comment (jkunle) [reply infrastructure/lambda/template.yaml:146]: @CharlesIC will it work, with a map?

Comment (mikebeeby) [reply infrastructure/lambda/template.yaml]: Done, although I did have to put the language extension after the Serverless transform or it did not work ` Template format error: Unresolved resource dependencies [PublicKBVApiStage] in the Resources block of the template
`
I also refactored it as we discussed so it is no longer defined for each env.

File: lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java [hunk 5]
```diff
@@ -78,6 +82,11 @@
 
 @ExtendWith(MockitoExtension.class)
 class QuestionHandlerTest {
+    private static final String MOCK_IIQ_STRATEGY_PARAM_VALUE =
+            "\"2\": \"3 out of 4 prioritised\"}";
```
Comment (jkunle) [inline lambdas/question/src/test/java/uk.gov.di.ipv.cri.kbv.api.handler/QuestionHandlerTest.java:86]: Looks like you missing "{"

File: lambdas/question/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionHandler.java [hunk 6]
```diff
@@ -249,6 +252,22 @@ private QuestionRequest createQuestionRequest(PersonIdentityDetailed personIdent
         return questionRequest;
     }
 
+    private String retrieveQuestionStrategy()
+            throws JsonProcessingException, InvalidStrategyScoreException {
+        var verificationScore = "2";
```
Comment (jkunle) [inline lambdas/question/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionHandler.java:257]: Maybe make this a constant

Comment (jkunle) [reply lambdas/question/src/main/java/uk/gov/di/ipv/cri/kbv/api/handler/QuestionHandler.java:257]: Maybe, could do something like `String strategy = strategyMap.getOrDefault(sessionVerificationScore. verificationScore);` will be a placeholder for the when we have something coming from the sessionItem

**Review by jkunle — APPROVED** [review]

**Pull request merged**